src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,357 @@
+                listeners?.Remove(listener);
+            });
+        }
+
+        internal void NotifyActivityStart(Activity activity)
+        {
+            Debug.Assert(activity != null);
+
+            // _listeners can get assigned to null in Dispose.
+            SynchronizedList<ActivityListener>? listeners = _listeners;
+            if (listeners != null &&  listeners.Count > 0)
+            {
+                listeners.EnumWithAction(listener => listener.ActivityStarted?.Invoke(activity));
+            }
+        }
+
+        internal void NotifyActivityStop(Activity activity)
+        {
+            Debug.Assert(activity != null);
+
+            // _listeners can get assigned to null in Dispose.
+            SynchronizedList<ActivityListener>? listeners = _listeners;
+            if (listeners != null &&  listeners.Count > 0)
+            {
+                listeners.EnumWithAction(listener => {
+                    listeners.EnumWithAction(listener => listener.ActivityStopped?.Invoke(activity));
+                });
+            }
+        }
+    }
+
+    // SynchronizedList<T> is a helper collection which ensure thread safety on the collection
+    // and allow enumerating the collection items and execute some action on the enumerated item and can detect any change in the collection
+    // during the enumeration which force restarting the enumeration again.
+    // Causion: We can have the action executed on the same item more than once which is ok in our scenarios.
+    internal class SynchronizedList<T>
+    {
+        private List<T> _list;
+        private uint _version;
+
+        public SynchronizedList() => _list = new List<T>();
+
+        public void Add(T item)
+        {
+            lock (_list)
+            {
+                _list.Add(item);
+                _version++;
+            }
+        }
+
+        public bool AddIfNotExist(T item)
+        {
+            lock (_list)
+            {
+                if (!_list.Contains(item))
+                {
+                    _list.Add(item);
+                    _version++;
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        public bool Remove(T item)
+        {
+            lock (_list)
+            {
+                if (_list.Remove(item))
+                {
+                    _version++;
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        public int Count => _list.Count;
+
+        public void EnumWithFunc(Func<T, bool> func)
+        {
+            uint version = _version;
+            int index = 0;
+
+            while (index < _list.Count)
+            {
+                T item;
+                lock (_list)
+                {
+                    if (version != _version)
+                    {
+                        version = _version;
+                        index = 0;
+                        continue;
```

Comment:
So if there's a mutation we start over? Hence the warning about potentially invoking delegates multiple times?  This means if one bad actor was repeatedly adding / removing / adding / removing / etc. while someone else was enumerating, that someone else might never end?

Reply:
Yes, it is safe in the calling scenarios. We are calling the listeners to detect if we need to create the Activity. I am not seeing a problem calling the listener multiple times with the same parameters multiple parameters when someone else mutates the list. 
The lists we have are the ActivtySource lists and listeners inside the activity source. If there is a bad actor you can do many more bad things creating a lot of Activity sources and listeners. or even in the listening can spend a long time in the execution.